### ISSUE TITLE
Wrong coverage after update Kotlin to 1.5.10

### ISSUE THREAD

**Issue by EduardoDornel** [jacoco/jacoco#1201]
Versions:
* JaCoCo version:  0.8.7.202105040129
* Operating system: macOS Mojave 10.14.6
* Tool integration: Gradle

After update kotlin from 1.4.32 to 1.5.10 the branch coverage has decreased

Coverage with Kotlin 1.4.32
![kotlin-1 4 32](https://user-images.githubusercontent.com/19539185/122577668-7039ba00-d029-11eb-84a6-005f86e11cdc.png)

Coverage with Kotlin 1.5.10
![kotlin-1 5 10](https://user-images.githubusercontent.com/19539185/122577772-8b0c2e80-d029-11eb-93a0-16357475f943.png)

* Kotlin Version: 1.5.10
* Gradle wrapper: 7.1
* Gradle plugin: 4.2.1
* Build tools: 30.0.2

**Comment by Godin** [comment]
@EduardoDornel Not much can be concluded or done on our side by just looking on your total numbers.
What if maybe previous numbers were actually incorrect? :wink:

So could you please provide concrete examples of cases?

**Comment by EduardoDornel** [comment]
A Fragment class with Kotlin 1.4.32

![kotlin 1 4 32-paymentReceipt](https://user-images.githubusercontent.com/19539185/122598898-d92e2b80-d043-11eb-82da-06ade97d61a4.png)

The same class with Kotlin 1.5.10

![kotlin 1 5 10-paymentReceipt](https://user-images.githubusercontent.com/19539185/122598915-e21efd00-d043-11eb-8f28-53b24a7fc92a.png)

This lines don't have "branches", they are only variable assignments

**Comment by Godin** [comment]
@EduardoDornel JaCoCo performs analysis of bytecode in class-files, so just by looking at partial screenshots unlikely that somebody will be able to guess why Kotlin compiler adds branches and what we can do with this. Hence please provide **complete example** that we'll be able compile by ourselves to reproduce this.

**Comment by Rolf-Smit** [comment]
I think I have an example (and reproducible project). I'm the author of [this](https://github.com/NeoTech-Software/Android-Root-Coverage-Plugin) Gradle Plugin for Android that automatically configures the Gradle JaCoCo plugin for use with multi-module Android projects.

My plugin comes with a integration test that verifies whether everything that should be covered is covered:

**Running those with Kotlin 1.4.31 and JaCoCo 0.8.7:**
![Kotlin 1.4 and JaCoCo 0.8.7](https://user-images.githubusercontent.com/6614322/123613881-1d6bb980-d804-11eb-87da-5124a99c19dd.png)

**However when updating the Kotlin version to 1.5.0, 1.5.10 or 1.5.20 that is used by integration test (test project):**
![Kotlin 1.5 and JaCoCo 8.7](https://user-images.githubusercontent.com/6614322/123614141-5b68dd80-d804-11eb-8f93-6d05715a7125.png)


**Context:**

- [LibraryAndroidKotlin.kt](https://github.com/NeoTech-Software/Android-Root-Coverage-Plugin/blob/feature/android-build-tools-7-and-kotlin-1.5/plugin/src/test/test-fixtures/multi-module/library_android/src/main/java/org/neotech/library/android/LibraryAndroidKotlin.kt) (the Kotlin file that fails to cover 100% on Kotlin 1.5)
- [Branch with Kotlin 1.4 that covers everything 100%](https://github.com/NeoTech-Software/Android-Root-Coverage-Plugin/tree/feature/android-build-tools-7-and-kotlin-1.4)
- [Branch with Kotlin 1.5 that fails to cover LibraryAndroidKotlin.kt](https://github.com/NeoTech-Software/Android-Root-Coverage-Plugin/tree/feature/android-build-tools-7-and-kotlin-1.5)
- It seems only the local JUnit coverage is missing, instrumented coverage (on an Android device works). Maybe this has something to do with the JVM version?

**Comment by Godin** [comment]
@Rolf-Smit are you sure that in your case required tests are really executed?

JaCoCo processes `touchedByJavaInstrumentedTestInAndroidLibrary` and `touchedByJavaUnitTestInAndroidLibrary` in the same way, so yes - because of 

> It seems only the local unit test coverage is missing, instrumented coverage (on an Android device) works

your issue does not seem related to JaCoCo and likely related to the way of execution of tests or something else.
Also because for the following `Example.kt` similar to yours `LibraryAndroidKotlin.kt`

```
object LibraryAndroidKotlin {

    @JvmStatic
    fun touchedByJavaInstrumentedTestInAndroidLibrary(): String {
        return "touchedByJavaInstrumentedTestInAndroidLibrary"
    }

    @JvmStatic
    fun touchedByJavaUnitTestInAndroidLibrary(): String {
        return "touchedByJavaUnitTestInAndroidLibrary"
    }

    fun touchedByKotlinInstrumentedTestInAndroidLibrary(): String {
        return "touchedByKotlinInstrumentedTestInAndroidLibrary"
    }

    fun touchedByKotlinUnitTestInAndroidLibrary(): String {
        return "touchedByKotlinUnitTestInAndroidLibrary"
    }

    @JvmStatic
    fun touchedByJavaInstrumentedTestInApp(): String {
        return "touchedByJavaInstrumentedTestInApp"
    }

    @JvmStatic
    fun touchedByJavaUnitTestInApp(): String {
        return "touchedByJavaUnitTestInApp"
    }

    fun touchedByKotlinInstrumentedTestInApp(): String {
        return "touchedByKotlinInstrumentedTestInApp"
    }

    fun touchedByKotlinUnitTestInApp(): String {
        return "touchedByKotlinUnitTestInApp"
    }
}

fun main() {
    LibraryAndroidKotlin.touchedByJavaInstrumentedTestInAndroidLibrary()
    LibraryAndroidKotlin.touchedByJavaUnitTestInAndroidLibrary()
    LibraryAndroidKotlin.touchedByKotlinInstrumentedTestInAndroidLibrary()
    LibraryAndroidKotlin.touchedByKotlinUnitTestInAndroidLibrary()
    LibraryAndroidKotlin.touchedByJavaInstrumentedTestInApp()
    LibraryAndroidKotlin.touchedByJavaUnitTestInApp()
    LibraryAndroidKotlin.touchedByKotlinInstrumentedTestInApp()
    LibraryAndroidKotlin.touchedByKotlinUnitTestInApp()
}
```

execution of

```
./kotlin-1.5.10/bin/kotlinc Example.kt -d classes
java -javaagent:jacoco-0.8.7/lib/jacocoagent.jar -cp classes ExampleKt
java -jar jacoco-0.8.7/lib/jacococli.jar report jacoco.exec --classfiles classes --sourcefiles . --html report
```

produces following report

<img width="1166" alt="Screenshot 2021-06-28 at 12 17 47" src="https://user-images.githubusercontent.com/138671/123620791-e220b900-d80a-11eb-9880-02f2174a64f6.png">

----

> Without emulator:
> Checkout the project (one of the 2 branches)

```
git clone https://github.com/NeoTech-Software/Android-Root-Coverage-Plugin.git
cd Android-Root-Coverage-Plugin
git checkout feature/android-build-tools-7-and-kotlin-1.5
```

> Change executeTests true to executeAndroidTests false here
> Change includeAndroidTestResults from true to false here

```diff --git i/plugin/src/test/test-fixtures/multi-module/build.gradle w/plugin/src/test/test-fixtures/multi-module/build.gradle
index 0706198..712d489 100644
--- i/plugin/src/test/test-fixtures/multi-module/build.gradle
+++ w/plugin/src/test/test-fixtures/multi-module/build.gradle
@@ -47,8 +47,8 @@ rootCoverage {
     generateXml false
     generateCsv true
     buildVariant "debug"
-    executeTests true
+    executeTests false
     includeUnitTestResults true
-    includeAndroidTestResults true
+    includeAndroidTestResults false
     includeNoLocationClasses true
 }
```

> Run ./gradlew test in the root of the project

```
org.neotech.plugin.rootcoverage.IntegrationTest > execute[multi-module-7.0.2] FAILED
    java.lang.IllegalStateException: Missing 'ANDROID_HOME' environment variable or local.properties with 'sdk.dir'
        at org.neotech.plugin.rootcoverage.util.ProjectGenerationKt.androidHomeDirectory(ProjectGeneration.kt:43)
        at org.neotech.plugin.rootcoverage.util.ProjectGenerationKt.createLocalPropertiesFile(ProjectGeneration.kt:11)
        at org.neotech.plugin.rootcoverage.util.ProjectGenerationKt.createLocalPropertiesFile$default(ProjectGeneration.kt:10)
        at org.neotech.plugin.rootcoverage.IntegrationTest.execute(IntegrationTest.kt:27)
```

> With emulator:
> Start an emulator

Sorry, but IMO it is strange to require heavy installation of Android SDK to demonstrate issue that is just about Kotlin version. Please understand that our resources are limited, so better to provide reduced examples.

**Comment by Rolf-Smit** [comment]
@Godin thanks for spending your time on this. I have to admit it is not the best reproducible example. I will try to see if I can find out whether or not the tests actually run. I really didn't think about that when I was up and downgrading the Kotlin version.

Edit:
- I found that the tests actually do run (added some std out lines, and I see them being printed)
- I found that when running with the coverage agent from within IntelliJ (as I described here: https://github.com/NeoTech-Software/Android-Root-Coverage-Plugin/issues/36#issuecomment-862259673) also does not report coverage. This rules out my plugin and also the JaCoCo Gradle Plugin, since the JaCoCo agent from IntelliJ itself is used (correct me if that is a wrong assumption).

I will try to make a minimal example that can reproduce this (preferably without any Android stuff). If it turns out that I can't manage to reproduce this without Android, I guess the problem is somewhere in the Android build tools, but we will see.

**Comment by Godin** [comment]
@Rolf-Smit 

> If it turns out that I can't manage to reproduce this without Android, I guess the problem is somewhere in the Android build tools

I highly suspect that your case is indeed related to Android SDK. Maybe https://github.com/NeoTech-Software/Android-Root-Coverage-Plugin/commit/ce49bef13f761ca363ccf6950fd35ce176aa567d

```diff --git a/plugin/src/test/test-fixtures/multi-module/build.gradle b/plugin/src/test/test-fixtures/multi-module/build.gradle
index c7c2b34..0706198 100644
--- a/plugin/src/test/test-fixtures/multi-module/build.gradle
+++ b/plugin/src/test/test-fixtures/multi-module/build.gradle
@@ -39,7 +39,7 @@ task clean(type: Delete) {
 
 jacoco {
     // https://www.jacoco.org/jacoco/trunk/doc/changes.html
-    toolVersion = "0.8.6"
+    toolVersion = "0.8.7"
 }
 
 rootCoverage {
```

is not enough in case of Android - see https://developer.android.com/reference/tools/gradle-api/7.1/com/android/build/api/dsl/CommonExtension?hl=he#testCoverage(kotlin.Function1)

> I will try to make a minimal example that can reproduce this (preferably without any Android stuff).

In absence of reproducer from @EduardoDornel weird to consider that he ran into the same problem as you, so let's stop mixing them - please open separate ticket if you can show problem with JaCoCo and Kotlin 1.5 unrelated to Android SDK.

Thank you for your understanding and good luck with investigations.

----

@EduardoDornel in absence of reproducer from you and in order to not attract unrelated discussions and comments, I'm closing this ticket, feel free to reopen with reproducer.

**Comment by paulnunezm** [comment]
I'm seeing the same issue after updating a multi module project to Kotlin 1.5.10 and AGP. My coverage went down more than 20%

**Comment by marchof** [comment]
@paulnunezm As @Godin said before: We can only work on this, if you provide a minimal reproducer. Please create a reproducer (preferrably Github repository) and open a new issue. Thx!

**Comment by JiongBull** [comment]
hi, i have the same issue, nothing is changed but kotlin version.
sorry, i have no project to submit currently, i will submit it later.

![1 4 32](https://user-images.githubusercontent.com/4413502/128959755-d8973675-2cf0-4fa9-9c8e-7961149ccf8c.png)

![1 5 21](https://user-images.githubusercontent.com/4413502/128959760-afc8bcc6-1e82-4b93-b8c0-6faabe31da82.png)

**Comment by JiongBull** [comment]
This is a simple project ([jacoco-kotlin](https://github.com/JiongBull/jacoco-kotlin)) which contains two modules, `core` and `app`. 

I found that jacoco won't make test coverage for unit tests in library module (`core`), but works for unit tests in apk module (`app`)

report for core module 👎 
![report for core](https://user-images.githubusercontent.com/4413502/128980858-3fe96a4c-905d-4706-847d-fb846f6a5618.png)

---

report for app module 👍 
![report for app](https://user-images.githubusercontent.com/4413502/128980907-d45d8453-a406-4dc0-9745-fe3fee9826ac.png)

**Comment by daniel00k** [comment]
@JiongBull the same is happening to me, but I realize that the coverage works again if you change your module plugin from 'com.android.library' to 'com.android.application'

kotlin version : "1.5.30"
jacoco version : "0.8.7"
gradle version : "6.7.1"

**Comment by meladRaouf** [comment]
> the same is happening to me, the coverage works again if Ichange module plugin from 'com.android.library' to 'com.android.application'

**Comment by minaEweida** [comment]
I got the same exact issue that tests in my android application module were not included in jacoco report after upgrading AGP 7 and kotlin 1.5.31 but in my case it was related to AGP 7 not kotlin.

And then I found out that the test execution location changed in the application module to something like

`YOUR_APP_MODULE/build/outputs/unit_test_code_coverage/(YOUR_PRODUCT_FLAVOR + YOUR_BUILD_TYPE)UnitTest/YOUR_EXEC_FILE.exec`

After updating the path to execution for my android application module (just added condition if current subproject is android application then I put the path above) in my coverage script and my coverage report was fixed.

**Comment by asthagarg2428** [comment]
Can anyone help me with https://stackoverflow.com/q/70151110/8762338.
Can't find any relevant solution

**Comment by sujeetsingh340** [comment]
@minaEweida  What is your below versions? I have changed same but still I can see code coverage is 0%.
kotlin version : "1.5.0"
jacoco version : "0.8.7"
gradle version : "7.0.2"

![image](https://user-images.githubusercontent.com/19792438/154034708-cd6e090f-9afd-4cc8-88db-3bb1ffb8d1f7.png)

**Comment by Rolf-Smit** [comment]
@sujeetsingh340 the problem is in your case AGP version 7.0.2. Because it has a bug causing execution files to be empty in some cases. This has nothing to do with JaCoCo.

I discovered this but myself for the first time, when the tests that verify that my own coverage plugin is working correctly started to fail after updating to AGP version 7.0. Since then I have not been able to release a stable version of my plugin, since this has not been fixed in any stable AGP version. You can read more about it in "note 2" here: https://github.com/NeoTech-Software/Android-Root-Coverage-Plugin#4-compatibility

_Or just go to the bug report directly: https://issuetracker.google.com/issues/195860510#comment12_


**TLDR: This issues is only fixed starting with AGP version 7.2-beta01 and has been present since AGP version 7.0**

**Comment by sujeetsingh340** [comment]
> @sujeetsingh340 the problem is in your case AGP version 7.0.2. Because it has a bug causing execution files to be empty in some cases. This has nothing to do with JaCoCo.
> 
> I discovered this but myself for the first time, when the tests that verify that my own coverage plugin is working correctly started to fail after updating to AGP version 7.0. Since then I have not been able to release a stable version of my plugin, since this has not been fixed in any stable AGP version. You can read more about it in "note 2" here: https://github.com/NeoTech-Software/Android-Root-Coverage-Plugin#4-compatibility
> 
> _Or just go to the bug report directly: https://issuetracker.google.com/issues/195860510#comment12_
> 
> **TLDR: This issues is only fixed starting with AGP version 7.2-beta01 and has been present since AGP version 7.0**
@Rolf-Smit  So If I use AGP version lower than 7.0.2 (ex 6.7.1) then it should work fine correct?
